### ISSUE TITLE
Telemetry

### ISSUE THREAD

**Issue by david-driscoll** [OmniSharp/OmniSharp#5]
## Telemetry
- Super controversial?
- Should we do it, should we not do it?
- _What if I don't want my data going to Microsoft?_

The answers were basically:
- **Must** be opt-in, ask the user, allow them to opt back out.
- The data should be open for all, this is open source after all
- The Telemetry isn't Microsoft telemetry, it's OmniSharp's Telemetry.

Otherbits:
- Possible Microsoft support to deal with telemetry, app insights or w/e? 

cc @OmniSharp/all

**Comment by sayedihashimi** [comment]
Thanks @david-driscoll for posting this.

When I brought up the idea of telemetry it wasn't meant to be to a repository that only Microsoft has access to :)
Ideally I'd like to have the info to be as public as possible.

While using `yeoman` and `bower` I noticed that they prompt for telemetry as well. Does anyone know where that data goes and how community members can access it? If it's public maybe we can copy what they are doing. If we don't have access to it or cannot figure out what they are doing, then maybe we can create a shared google analytics account or make the data public in the form of a drop on some regular basis.

The idea behind adding telemetry is to enable us to **_make informed decisions**_ instead of guessing on what components are getting used.

**Comment by luisrudge** [comment]
Maybe create a custom dashboard using google analytics api?

**Comment by dmccaffery** [comment]
I completely agree that telemetry would be hugely beneficial to the community; it will allow us to focus on what is most important, as well as to understand where our contributions are taking hold. Not to mention that it would act as a strong motivator to continue development when usage proves that it's useful.

**Comment by ctolkien** [comment]
There's a bit of negative dev mindshare going around at the moment with regard to MS and tracking, so I think that's where the hesitation comes from.

With that out of the way, bower do a good job of exposing this sort of high level info:
http://bower.io/stats/

The info goes into Google Analytics.

Further discussion from bower:
https://github.com/bower/bower/issues/1164

and Yeoman:
https://github.com/yeoman/yeoman/issues/470

And their 'Insight' bits:
https://github.com/yeoman/insight
and also their dashboard (which I can't see online anywhere at the moment):
https://github.com/yeoman/insight-dashboard

**Comment by mikavilpas** [comment]
For Emacs we have the more humble download counter at
https://melpa.org/#/omnisharp

But from the meeting I got the idea that editors should report usage
statistics when the user is installing/starting the plugin. Is this what
you meant?

I would love more accurate statistics, as long as the users comply with
giving that information.

2015-12-03 1:29 GMT+02:00 Chad T notifications@github.com:

**Comment by david-driscoll** [comment]
At a high level I'd like to see...
- Editor plugin and version (or commit sha?)
- OmniSharp type (server / roslyn) and version (or commit sha?)
- Session time?

At a lower level it might be nice to know what api's editors are using, and what of those API's users are using in a typical session, we probably don't need to stream those real time though, maybe just cache to a file or something and synchronize that periodically.

**Comment by luisrudge** [comment]
I think that bad telemetry is worse than no telemetry.. So, first, I think it's best to ask the community through the official channels telemetry is something the're willing to share.

**Comment by jrieken** [comment]
Also, interesting to know
- project type (msbuild, dnx, script cs)
- usage frequency of what features
- perf behaviour

**Comment by david-driscoll** [comment]
@luisrudge what official channels? lol... I guess if someone gets me the OmniSharp twitter account info I could ask there. :smile:

**Comment by sayedihashimi** [comment]
@jchannon can you post a link to this from the OmniSharp twitter account?

> So, first, I think it's best to ask the community through the official channels telemetry is something the're willing to share.

@luisrudge that's the purpose of this issue. What "official channels" are you referring to? We discussed the topic in our last online meeting and we have posted this here to get community feedback. I think that is as "official" as it gets.

**Comment by jchannon** [comment]
I've lost the password

@mat-mcloughlin or @Mpdreamz or @david-driscoll might have it

**Comment by david-driscoll** [comment]
I don't have it! lol

**Comment by sayedihashimi** [comment]
Who created the twitter account?

**Comment by jchannon** [comment]
@shanselman

**Comment by dmccaffery** [comment]
The least Scott! How dare he.

**Comment by luisrudge** [comment]
@sayedihashimi the problem with posting here or the online meeting is that we represent what? 1% of the users? We're interested in making it better, so everyone will say "oh yeah, I'd like some telemetry". If we take that as the truth, we might get bitten by the real world usage, which may or may not want telemetry at all.

**Comment by dmccaffery** [comment]
@luisrudge ... so we bite back harder. :) I kid, I kid... I think that the only viable solution to getting a feel from the real world as to whether or not they will accept telemetry is to start asking them directly after they download or install an OmniSharp package/tool/whatever.

In reality, those not involved in building OmniSharp products or solutions aren't going to WANT telemetry, but they might ACCEPT it. The only benefit to them is that more time and energy will go into the functionality they want most, which is a tough sell to someone that just wants to get something done right now.

The benefit to us, of course, is that we spend our limited time focusing on what matters most. Best bang for the buck, so to speak.

I'm all for an optin scenario, but I don't think we'll ever get a reasonable feel for how well the community accepts telemetry until we actually try it.

**Comment by luisrudge** [comment]
> but I don't think we'll ever get a reasonable feel for how well the community accepts telemetry until we actually try it.

Fully agree!
